std/arith.asm

```diff
@@ -17,18 +23,176 @@ machine Arith(CLK32_31, operation_id){
+    let s_for_eq2 = |x1, y1, limb_index| (div(mul(3, mul(x1, x1)), mul(2, y1)) >> (limb_index * 16)) & 0xffff;
+
+    let compute_x3_int = |x1, x2, s| (s * s - x1 - x2 + 2 * secp_modulus) % secp_modulus;
+    let compute_y3_int = |x1, y1, x3, s| (s * (((x1 - x3) + secp_modulus) % secp_modulus) - y1 + secp_modulus) % secp_modulus;
+
+    // Note that the most significant limb is 32-Bit instead of 16
+    let compute_q0_for_eq1 = |x1, y1, x2, y2, s, limb_index| ((-(s * x2 - s * x1 - y2 + y1) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+    let compute_q0_for_eq2 = |x1, y1, s, limb_index| ((-(2 * s * y1 - 3 * x1 * x1) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+    let compute_q1 = |x1, x2, x3, s, limb_index| ((-(s * s - x1 - x2 - x3) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+    let compute_q2 = |x1, y1, x3, y3, s, limb_index| ((-(s * x1 - s * x3 - y1 - y3) / secp_modulus + (1 << 258)) >> (limb_index * 16)) & if limb_index < 15 { 0xffff } else { 0xffffffff };
+ 
+    let limbs_to_int = |limbs, row| array::sum(array::new(array::len(limbs), |i| int(limbs[i](row)) << (i * 16)));
+
+    let x1_int: int -> int = |row| limbs_to_int(x1, row);
+    let y1_int: int -> int = |row| limbs_to_int(y1, row);
+    let x2_int: int -> int = |row| limbs_to_int(x2, row);
+    let y2_int: int -> int = |row| limbs_to_int(y2, row);
+    let x3_int: int -> int = |row| limbs_to_int(x3, row);
+    let s_int: int -> int = |row| limbs_to_int(s, row);
+
+    let s_hint = |i, limb_index| if selEq[1](i) == fe(1) {
+        s_for_eq1(x1_int(i), y1_int(i), x2_int(i), y2_int(i), limb_index)
+    } else {
+        if selEq[2](i) == fe(1) {
+            s_for_eq2(x1_int(i), y1_int(i), limb_index)
+        } else {
+            0
+        }
+    };
+
+    let q0_hint = |i, limb_index| if selEq[1](i) == fe(1) {
+        compute_q0_for_eq1(x1_int(i), y1_int(i), x2_int(i), y2_int(i), s_int(i), limb_index)
+    } else {
+        if selEq[2](i) == fe(1) {
+            compute_q0_for_eq2(x1_int(i), y1_int(i), s_int(i), limb_index)
+        } else {
+            0
+        }
+    };
+
+    let q1_hint = |i, limb_index| if selEq[3](i) == fe(1) {
+        // TODO: Make this more readable once we have let statements
+        (|x1, x2, s|
+            compute_q1(x1, x2, compute_x3_int(x1, x2, s), s, limb_index)
+        )(x1_int(i), x2_int(i), s_int(i))
+    } else {
+        0
+    };
+
+    let q2_hint = |i, limb_index| if selEq[3](i) == fe(1) {
+        // TODO: Make this more readable once we have let statements
+        (|x1, y1, x2, s|
+            (|x3|
+                compute_q2(x1, y1, x3, compute_y3_int(x1, y1, x3, s), s, limb_index)
+            )(compute_x3_int(x1, x2, s))
+        )(x1_int(i), y1_int(i), x2_int(i), s_int(i))
```

Comment:
This used to be:
`compute_q2(x1_int(i), y1_int(i), compute_x3_int(x1_int(i), x2_int(i), s_int(i)), compute_y3_int(x1_int(i), y1_int(i), compute_x3_int(x1_int(i), x2_int(i), s_int(i)), s_int(i)), s_int(i), limb_index)`

This get's rid of duplicate calls to various functions. If we had let statements, the equivalent code would be:
```
let x1 = x1_int(i);
let y1 = y1_int(i);
let x2 = x2_int(i);
let s = s_int(i);
let x3 = compute_x3_int(x1, x2, s);
let y3 = compute_y3_int(x1, y1, x3, s);
compute_q2(x1, y1, x3, y3,, s, limb_index)
```